api/server/runner.go

```diff
@@ -102,16 +114,46 @@ func (s *Server) serve(c *gin.Context, appName, path string) error {
 			// add this, since it means that start may not have been called [and it's relevant]
 			c.Writer.Header().Add("XXX-FXLB-WAIT", time.Now().Sub(time.Time(model.CreatedAt)).String())
 		}
-		// NOTE: if the task wrote the headers already then this will fail to write
-		// a 5xx (and log about it to us) -- that's fine (nice, even!)
 		return err
 	}
 
-	// TODO plumb FXLB-WAIT somehow (api?)
+	// if they don't set a content-type - detect it
+	if writer.Header().Get("Content-Type") == "" {
+		// see http.DetectContentType, the go server is supposed to do this for us but doesn't appear to?
+		var contentType string
+		if bytes.HasPrefix(buf.Bytes(), jsonPrefix) {
+			// try to detect json, since DetectContentType isn't a hipster.
+			contentType = "application/json; charset=utf-8"
+		} else {
+			contentType = http.DetectContentType(buf.Bytes())
+		}
+		writer.Header().Set("Content-Type", contentType)
+	}
 
-	// TODO we need to watch the response writer and if no bytes written
-	// then write a 200 at this point?
-	// c.Data(http.StatusOK)
+	writer.Header().Set("Content-Length", strconv.Itoa(int(buf.Len())))
+
+	if writer.status > 0 {
+		c.Writer.WriteHeader(writer.status)
+	}
+	io.Copy(c.Writer, &writer)
 
 	return nil
 }
+
+var jsonPrefix = []byte("{")
+
+var _ http.ResponseWriter = new(syncResponseWriter)
```

Comment:
do we need this line?

Reply:
this is something of an idiom so that the compiler will generate a nice message that says we don't fulfill that interface directly here instead of generating error messages from the call sites where it is used. as a plus it nicely documents its use